tracer/src/Datadog.Tracer.Native/debugger_method_rewriter.cpp

```diff
@@ -758,6 +715,462 @@ HRESULT DebuggerMethodRewriter::ApplyMethodProbe(
+            return E_FAIL;
+        }
+
+        // get the interface type props
+        WCHAR type_name[kNameMaxSize]{};
+        DWORD type_name_len = 0;
+        mdAssembly assemblyToken;
+        if (metadataImport->GetTypeRefProps(interfaceToken, &assemblyToken, type_name, kNameMaxSize, &type_name_len) !=
+            S_OK)
+        {
+            Logger::Warn("DebuggerMethodRewriter::IsTypeImplementIAsyncStateMachine: failed to get type ref props");
+            return E_FAIL;
+        }
+
+        // if the interface is the IAsyncStateMachine
+        if (classToken == typeToken && type_name == debugger_iasync_state_machine_name)
+        {
+            return S_OK;
+        }
+
+        return S_FALSE;
+    }
+
+    return E_FAIL;
+}
+
+HRESULT DebuggerMethodRewriter::IsAsyncMethodProbe(const ComPtr<IMetaDataImport2>& metadataImport,
+                                                   const FunctionInfo* caller) const
+{
+    if (caller->name != WStr("MoveNext") || caller->method_signature.NumberOfArguments() > 0 ||
+        std::get<unsigned>(caller->method_signature.GetReturnValue().GetElementTypeAndFlags()) != ELEMENT_TYPE_VOID)
+    {
+        return S_FALSE;
+    }
+
+    return IsTypeImplementIAsyncStateMachine(metadataImport, caller->type.id);
+}
+
+HRESULT DebuggerMethodRewriter::GetTaskReturnType(const ILInstr* instruction, ModuleMetadata& moduleMetadata, const std::vector<TypeSignature>& methodLocals, TypeSignature* returnType)
+{
+    for (const ILInstr* pInstr = instruction->m_pPrev;
+        pInstr != instruction; 
+        pInstr = pInstr->m_pPrev)
+    {
+        // It is a call to a struct method so CALL instruction but pay attention to change it if the runtime changes
+        if (pInstr->m_opcode != CEE_CALL)
+        {
+            continue;
+        }
+
+        auto functionInfo = GetFunctionInfo(moduleMetadata.metadata_import, pInstr->m_Arg32);
+        if (functionInfo.name == WStr("SetException"))
+        {
+            // We go through the instructions in reverse order so if we're already in SetException,
+            // it means that there is not SetResult in the current MoveNext method
+            return S_FALSE;
+        }
+
+        if (functionInfo.name != WStr("SetResult"))
+        {
+            continue;
+        }
+
+        if (functionInfo.name == WStr("SetResult"))
+        {
+            if (ILRewriter::IsLoadLocalDirectInstruction(pInstr->m_pPrev->m_opcode) /*meaning that the task return T value*/)
+            {
+                // get the index of the local that represent the return value of the task
+                const auto returnValueLocalIndex =
+                    pInstr->m_pPrev->m_opcode == CEE_LDLOC
+                        ? pInstr->m_pPrev->m_Arg16
+                    : pInstr->m_pPrev->m_opcode == CEE_LDLOC_S
+                        ? pInstr->m_pPrev->m_Arg8
+                        : pInstr->m_pPrev->m_opcode - 6 /*6 because all the shortcuts to ldloc is the local index + 6*/;
```

Comment:
It should not break if we will add more locals, because anyway I try to get here index to original local and not something that we added. The condition result will change depends of the opcode but the logic should handle all cases. Anyway, I extracted this to new method `GetLocalIndexFromOpcode` in 1fa418a83881f1af240463893db4ee34fb543c45